12-offer-encoding.md

```diff
@@ -0,0 +1,1053 @@
+          - MUST set `proportional_amount` to 1 
+        - otherwise:
+          - MUST set `proportional_amount` to 0
+  - otherwise:
+    - MUST NOT include `recurrence_base`.
+    - MUST NOT include `recurrence_paywindow`.
+    - MUST NOT include `recurrence_limit`.
+  - if `send_invoice` is present:
+    - if the offer is for a partial or full refund for a previously-paid
+      invoice:
+      - SHOULD set `refunded_payment_hash` to the `payment_hash` of that
+        invoice.
+  - otherwise:
+    - MUST NOT set `refunded_payment_hash`.
+
+A reader of an offer:
+  - if `features` contains unknown _odd_ bits that are non-zero:
+    - MUST ignore the bit.
+  - if `features` contains unknown _even_ bits that are non-zero:
+    - MUST NOT respond to the offer.
+    - SHOULD indicate the unknown bit to the user.
+  - if `node_id` or `description` is not set:
+    - MUST NOT respond to the offer.
+  - if `description` is not present:
+    - MUST NOT respond to the offer.
+  - if `signature` is present, but is not a valid signature using `node_id` as described in [Signature Calculation](#signature-calculation):
+    - MUST NOT respond to the offer.
+  - SHOULD gain user consent for recurring payments.
+  - SHOULD allow user to view and cancel recurring payments.
+  - if it uses `amount` to provide the user with a cost estimate:
+    - MUST warn user if amount of actual invoice differs significantly
+        from that expectation.
+  - SHOULD not respond to an offer if the current time is after
+    `absolute_expiry`.
+  - FIXME: more!
+
+## Rationale
+
+It's quite reasonable to set a `recurrence_paywindow` with seconds_after
+equal to 0, but obviously this should not apply to the initial period if
+there is no recurrence_base.
+
+A signature is optional, because it makes for a longer string (potentially
+limiting QR code use on low-end cameras); if the offer has an error, no
+invoice will be given (or, for `send_invoice` offers, accepted), since
+the `offer_id` already covers all the non-signature fields.
+
+# Invoice Requests
+
+Invoice Requests are a request for an invoice; the human-readable prefix for
+invoices is `lnr`.
+
+## TLV Fields for `invoice_request`
+
+1. `tlv_stream`: `invoice_request`
+2. types:
+    1. type: 2 (`chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 4 (`offer_id`)
+    2. data:
+        * [`sha256`:`offer_id`]
+    1. type: 8 (`amount`)
+    2. data:
+        * [`tu64`:`msat`]
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 32 (`quantity`)
+    2. data:
+        * [`tu64`:`quantity`]
+    1. type: 36 (`recurrence_counter`)
+    2. data:
+        * [`tu32`:`counter`]
+    1. type: 68 (`recurrence_start`)
+    2. data:
+        * [`tu32`:`period_offset`]
+    1. type: 38 (`payer_key`)
+    2. data:
+        * [`point32`:`key`]
+    1. type: 39 (`payer_note`)
+    2. data:
+        * [`...*utf8`:`note`]
+    1. type: 50 (`payer_info`)
```

Comment:
Probably a stupid question and I should probably know the answer, but how big can this field be?

Reply:
No well-defined limit.  The receiver is free to ignore if it wants, too.  (In practice, c-lightning won't create an onion > 32768 bytes, so don't get too ambitious!).

Reply:
Oh, ok, that's awesome. If some design needs more than that there's probably something wrong with whatever design needs more than that.
I mean, you can always chain messages to make a total bigger message.
I think.

Reply:
Awesome. That's more than I should need for my use case.